vm/streaming.go

```diff
@@ -280,185 +77,86 @@ func (d *DecisionRPCClient) Close() error {
+	defer span.End()
+	// Unmarshal TX
+	p := codec.NewReader(msgBytes, chain.NetworkSizeLimit) // will likely be much smaller
+	tx, err := chain.UnmarshalTx(p, vm.actionRegistry, vm.authRegistry)
+	if err != nil {
+		vm.snowCtx.Log.Error("failed to unmarshal tx",
+			zap.Int("len", len(msgBytes)),
+			zap.Error(err),
+		)
 
-	// Wait for VM to be ready before accepting connections. If we stop the VM
-	// before this happens, we should return.
-	if !r.vm.waitReady() {
 		return
 	}
-
-	for {
-		nconn, err := r.listener.Accept()
-		if err != nil {
-			r.vm.snowCtx.Log.Warn("unable to accept connection", zap.Error(err))
-			return
-		}
-
-		// Generate connection id
-		var id ids.ID
-		_, err = rand.Read(id[:])
-		if err != nil {
-			panic(err)
-		}
-
-		// Store id
-		c := make(chan *chain.StatelessBlock, r.vm.config.GetStreamingBacklogSize())
-		r.vm.listeners.AddBlockListener(id, c)
-		r.vm.snowCtx.Log.Info("created new block listener", zap.Stringer("id", id))
-
-		conn := &blockRPCConnection{
-			conn: nconn,
-
-			vm:   r.vm,
-			blks: c,
-			id:   id,
-		}
-
-		// Nothing to read
-		go conn.handleWrites()
+	// Verify tx
+	sigVerify := tx.AuthAsyncVerify()
+	if err := sigVerify(); err != nil {
+		vm.snowCtx.Log.Error("failed to verify sig",
+			zap.Error(err),
+		)
+		return
 	}
-}
-
-func (r *blockRPCServer) Close() error {
-	return r.listener.Close()
-}
-
-func (c *blockRPCConnection) handleWrites() {
-	defer func() {
-		c.vm.metrics.blocksRPCConnections.Dec()
-		_ = c.conn.Close()
-		c.vm.listeners.RemoveBlockListener(c.id)
-	}()
+	// TODO: add tx associated with this connection
+	vm.listeners.AddTxListener(tx, c)
 
-	c.vm.metrics.blocksRPCConnections.Inc()
-	for {
-		select {
-		case blk := <-c.blks:
-			if err := writeNetMessage(c.conn, blk.Bytes()); err != nil {
-				c.vm.snowCtx.Log.Error("unable to send blk", zap.Error(err))
-				return
-			}
-			results, err := chain.MarshalResults(blk.Results())
-			if err != nil {
-				c.vm.snowCtx.Log.Error("unable to marshal blk results", zap.Error(err))
-				return
-			}
-			if err := writeNetMessage(c.conn, results); err != nil {
-				c.vm.snowCtx.Log.Error("unable to send blk results", zap.Error(err))
-				return
-			}
-		case <-c.vm.stop:
-			return
-		}
+	// Submit will remove from [txWaiters] if it is not added
+	txID := tx.ID()
+	if err := vm.Submit(ctx, false, []*chain.Transaction{tx})[0]; err != nil {
+		vm.snowCtx.Log.Error("failed to submit tx",
+			zap.Stringer("txID", txID),
+			zap.Error(err),
+		)
+		return
 	}
+	vm.snowCtx.Log.Debug("submitted tx", zap.Stringer("id", txID))
 }
 
-// If you don't keep up, you will data
 type BlockRPCClient struct {
-	conn net.Conn
+	conn *websocket.Conn
 
 	ll sync.Mutex
 	cl sync.Once
 }
 
+// NewBlockRPCClient creates a new client for the blocks rpc server.
+// Dials into the server at [uri] and returns a client.
 func NewBlockRPCClient(uri string) (*BlockRPCClient, error) {
```

Comment:
We should unify these clients I think. Seems odd to require a separate client/port for each function.